File: Assets/Scripts/Controller/EventManager.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using LitJson;

public class FishEvent
{
    public int et;  //eventtype
    public int t;   //time
    public int begint;  //begintime
    public int endt;  //endtime

    public FishEvent()
    { }
    public FishEvent(int eventtime, int time)
    {
        et = eventtime;
        t = time;
    }
}

public class EventManager
{
	private static EventManager mInstance = null;
    private EventManager()
    { }

    private List<FishEvent> mEventList = new List<FishEvent>();

    private FishEvent mEvent = null;

    private float timer = 2;

    public static EventManager GetInstance()
    {
        if (mInstance == null)
            mInstance = new EventManager();

        return mInstance;
    }

    public void Initialize()
    {
        mEventList = new List<FishEvent>();
        //mEventList.Add(new FishEvent(0, 60));
        //mEventList.Add(new FishEvent(1, 60));
        //string jsonStr = JsonMapper.ToJson(mEventList);
        //string filepath = Application.dataPath + "/Resources/event.bytes";
        //JsonUtil.Save(jsonStr, filepath);

        //LoadEventConfig();
        //CaculateBeginAndEndTime();
    }

    public void Update(float dt)
    {
        if (FishData.GetInstance().GameState == GameState.MainLoop)
        {
            timer += dt;
            if (timer > 3f)
            {
                timer = 0;
                TestTeam();
            }
        }
        return;
        mEvent = GetEvent(dt);
        if(mEvent == null) return;
        switch (mEvent.et)
        {
            case 0:
                break;
            case 1:
                break;
            case 2:
                break;
            default:
                break;
        }
    }

    public void LoadEventConfig()
    {
        string filepath = Application.dataPath + "/Resources/event.bytes";
        FileStream fs = new FileStream(filepath,FileMode.Open);
        StreamReader sr = new StreamReader(fs);
        string jsonStr = sr.ReadToEnd();
        mEventList = JsonMapper.ToObject<List<FishEvent>>(jsonStr);
    }

    public void CaculateBeginAndEndTime()
    {
        if (mEventList != null)
        {
            int t = 0;
            foreach (FishEvent eventdata in mEventList)
            {
                eventdata.begint = t;
                eventdata.endt = t + eventdata.t;
                t = eventdata.endt;
            }
        }
    }

    public FishEvent GetEvent(float t)
    {
        foreach (FishEvent eventdata in mEventList)
        {
            if (t < eventdata.endt)
                return eventdata;
        }
        return null;
    }

    public void TestTeam()
    {
        float hBound = 120;
        float yBottom = -70;
        float yUp = 70;
        int[] temp = new int[2] { -1, 1 };
        int flag = temp[Random.Range(0, 2)];
        float speed = Random.Range(40, 70);
        int pathid = Random.Range(0, 5);
        Vector3 headPosition = new Vector3(hBound * flag, Random.Range(yBottom + 20, yUp - 20), Random.Range(96, 96 + 20));
        //Vector3 bornEulerAngles = new Vector3(Random.Range(-20, 20), Random.Range(80, 100) * flag, 0);
        Vector3 bornEulerAngles = new Vector3(0, -90 * flag, 0);
        int randomCnt = Random.Range(5,11);
        for (int i = 0; i < randomCnt; i++)
        {
            Vector3 offset = new Vector3(i * 15 * flag, 0, 0);
            //offset = Quaternion.Euler(bornEulerAngles) * offset;
            FishManager.GetInstance().CreateFish(0, headPosition + offset, bornEulerAngles, pathid, speed, i * 15.0f / speed);
        }
    }
}

File: Assets/Framework/Lua/Logic/EventManager.lua
--region *.lua
--Date
--此文件由[BabeLua]插件自动生成
local json = require("cjson")
local EventManager = class("EventManager")

function EventManager:ctor()
    self.eventList_ = {}
    self.timer_ = 0
end

function EventManager:getInstance()
    if self.instance_ == nil then
        self.instance_ = EventManager.new()
    end

    return self.instance_;
end

function EventManager:loadEventConfig()
    local str = ResourceManager:LoadFile("config/tables","event")
    self.eventList_ = json.decode(str)
end

function EventManager:caculateBeginAndEndTime()
    if self.eventList_ ~= null then
        local t = 0
        for i,value in ipairs(self.eventList_) do
            value.beginTime = t
            value.endTime = t + value.t
            t = value.endTime
        end
    end
end

function EventManager:onUpdate(dt)
    self.timer_ = self.timer_ + dt
    if self.timer_ > 0.4 then
        self.timer_ = 0
        self:testOneFish()
    end
end

function EventManager:testOneFish()
    local hBound = 160;
    local yBottom = -70;
    local yUp = 70;
    local temp = { -1, 1 };
    local flag = temp[math.random(1, 2)];
    local speed = math.random(40, 70);
    local pathid = math.random(0, 5);
    local temp1 = { 0, 0, 0, 0, 0, 2, 0, 0, 0, 1 };
    local fishid = temp1[math.random(1, 10)];
    if fishid == 1 then
        speed = 30
    end
    if fishid == 2 then
        speed = 20
        pathid = 0
    end
    local headPosition = Vector3.New(hBound * flag, math.random(yBottom + 20, yUp - 20), math.random(150, 150 + 20));
    local bornEulerAngles = Vector3.New( math.random(-20, 20), - math.random(80, 100) * flag, 0);
    MainView:createFish(fishid, headPosition, bornEulerAngles, pathid, speed, 0);
end

return EventManager
--endregion
